src/gui/editlocallyhandler.cpp

```diff
@@ -0,0 +1,333 @@
+#include "folder.h"
+#include "folderman.h"
+#include "syncengine.h"
+#include "systray.h"
+
+namespace OCC {
+
+Q_LOGGING_CATEGORY(lcEditLocallyHandler, "nextcloud.gui.editlocallyhandler", QtInfoMsg)
+
+EditLocallyHandler::EditLocallyHandler(const QString &userId,
+                                       const QString &relPath,
+                                       const QString &token,
+                                       QObject *parent)
+    : QObject{parent}
+    , _userId(userId)
+    , _relPath(relPath)
+    , _token(token)
+{
+}
+
+void EditLocallyHandler::startSetup()
+{
+    if (_token.isEmpty() || _relPath.isEmpty() || _userId.isEmpty()) {
+        qCWarning(lcEditLocallyHandler) << "Could not start setup."
+                                        << "token:" << _token
+                                        << "relPath:" << _relPath
+                                        << "userId" << _userId;
+        return;
+    }
+
+    // Show the loading dialog but don't show the filename until we have
+    // verified the token
+    Systray::instance()->createEditFileLocallyLoadingDialog({});
+
+    // We check the input data locally first, without modifying any state or
+    // showing any potentially misleading data to the user
+    if (!isTokenValid(_token)) {
+        qCWarning(lcEditLocallyHandler) << "Edit locally request is missing a valid token, will not open file. "
+                                        << "Token received was:" << _token;
+        showError(tr("Invalid token received."), tr("Please try again."));
+        return;
+    }
+
+    if (!isRelPathValid(_relPath)) {
+        qCWarning(lcEditLocallyHandler) << "Provided relPath was:" << _relPath << "which is not canonical.";
+        showError(tr("Invalid file path was provided."), tr("Please try again."));
+        return;
+    }
+
+    _accountState = AccountManager::instance()->accountFromUserId(_userId);
+
+    if (!_accountState) {
+        qCWarning(lcEditLocallyHandler) << "Could not find an account " << _userId << " to edit file " << _relPath << " locally.";
+        showError(tr("Could not find an account for local editing."), tr("Please try again."));
+        return;
+    }
+
+    // We now ask the server to verify the token, before we again modify any
+    // state or look at local files
+    startTokenRemoteCheck();
+}
+
+void EditLocallyHandler::startTokenRemoteCheck()
+{
+    if (!_accountState || _relPath.isEmpty() || _token.isEmpty()) {
+        qCWarning(lcEditLocallyHandler) << "Could not start token check."
+                                        << "accountState:" << _accountState
+                                        << "relPath:" << _relPath
+                                        << "token:" << _token;
+        return;
+    }
+
+    const auto encodedToken = QString::fromUtf8(QUrl::toPercentEncoding(_token)); // Sanitise the token
+    const auto encodedRelPath = QUrl::toPercentEncoding(_relPath); // Sanitise the relPath
```

Comment:
Is this variable used anywhere?

Reply:
Sorry, messed something up during a prior rebase. Now using